OutOfSchool/OutOfSchool.WebApi/Services/SubordinationStructure/InstitutionHierarchyService.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+using OutOfSchool.Services.Models.SubordinationStructure;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.WebApi.Common;
+using OutOfSchool.WebApi.Models.SubordinationStructure;
+
+namespace OutOfSchool.WebApi.Services.SubordinationStructure
+{
+    public class InstitutionHierarchyService : IInstitutionHierarchyService
+    {
+        private readonly ISensitiveEntityRepository<InstitutionHierarchy> repository;
+        private readonly IWorkshopRepository repositoryWorkshop;
+        private readonly IProviderRepository repositoryProvider;
+        private readonly ILogger<InstitutionHierarchyService> logger;
+        private readonly IStringLocalizer<SharedResource> localizer;
+        private readonly IMapper mapper;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InstitutionHierarchyService"/> class.
+        /// </summary>
+        /// <param name="repository">Repository for InstitutionHierarchy entity.</param>
+        /// <param name="repositoryWorkshop">Workshop repository.</param>
+        /// <param name="repositoryProvider">Provider repository.</param>
+        /// <param name="logger">Logger.</param>
+        /// <param name="localizer">Localizer.</param>
+        /// <param name="mapper">Mapper.</param>
+        public InstitutionHierarchyService(
+            ISensitiveEntityRepository<InstitutionHierarchy> repository,
+            IWorkshopRepository repositoryWorkshop,
+            IProviderRepository repositoryProvider,
+            ILogger<InstitutionHierarchyService> logger,
+            IStringLocalizer<SharedResource> localizer,
+            IMapper mapper)
+        {
+            this.repository = repository ?? throw new ArgumentNullException(nameof(repository));
+            this.repositoryWorkshop = repositoryWorkshop;
```

Comment:
Please either validate all arguments using `?? throw new ArgumentNullException` in constructor or check them for `null` everywhere in code (like in line 53 you could get `NullReferenceException` if logger is null).

Reply:
Done